shared/src/main/scala/mlscript/NuTypeDefs.scala

```diff
@@ -985,6 +1060,54 @@ class NuTypeDefs extends ConstraintSolver { self: Typer =>
               
             case td: NuTypeDef =>
               
+              /** Check no `this` access in ctor statements or val rhs and reject unqualified accesses to virtual members.. */
+              def qualificationCheck(members: Ls[NuMember], stmts: Ls[Statement], baseAndSigs: Ls[NuMember]): Unit = {
+                val cache = mutable.HashMap[Str, Bool]()
+
+                def getMember(name: Str) =
+                  members.find(_.name === name).fold(baseAndSigs.find(_.name === name))(m => S(m))
+
+                // Return true if it is invalid
+                def checkThisInCtor(refs: RefMap, name: Opt[Str], stack: Ls[Str])(expection: Bool): Bool = {
+                  def run: Bool = {
+                    refs.useThis || (
+                      refs.refs.foldLeft(false)((res, p) => res || (getMember(p._1) match {
+                        case S(nf: TypedNuFun) if p._1 =/= name.getOrElse("") && !stack.contains(p._1) =>
+                          (p._2 && (!expection || nf.fd.isVirtual)) || checkThisInCtor(nf.getFunRefs, S(p._1), p._1 :: stack)(false)
+                        case _ => false // Refer to outer 
+                      }))
+                    )
+                  }
+
+                  name.fold(run)(name => cache.getOrElseUpdate(name, run))
+                }
+
+                def checkUnqualifiedVirtual(refs: RefMap, parentLoc: Opt[Loc]) =
+                  refs.refs.foreach(p => if (!p._2) getMember(p._1) match { // unqualified access
+                    case S(nf: TypedNuFun) if nf.fd.isVirtual =>
+                      err(msg"Unqualified access to virtual member ${p._1}" -> parentLoc ::
+                        msg"Declared here:" -> nf.fd.toLoc
+                      :: Nil)
+                    case _ => ()
+                  })
+
+                members.foreach {
+                  case tf @ TypedNuFun(_, fd, _) =>
+                    val refs = tf.getFunRefs
+                    if (fd.isLetRec.isDefined && checkThisInCtor(refs, S(tf.name), tf.name :: Nil)(true)) // not a function && access `this` in the ctor
+                      err(msg"Can not access `this` when initializing field ${tf.name}", fd.toLoc)
```

Comment:
The location here is wrong. Instead of returning a Boolean, `checkThisInCtor` should return the `Var` of the offending reference, so we can use its location.

Reply:
Updated `checkThisInCtor`. Now it returns `Option[Var]`: `Some(v: Var)` means there is an invalid access where the `v` indicates the corresponding reference, and `None` means the access is safe.